drv/lpc55-update-server/src/main.rs

```diff
@@ -573,6 +628,406 @@ impl ServerImpl<'_> {
+            };
+
+            indirect_flash_read(
+                &self.flash,
+                addr as u32,
+                &mut flash_page[..size],
+            )?;
+            // XXX fw_cache is [u32; FW_CACHE_MAX]
+            if flash_page[0..size] != cached[addr..addr + size] {
+                return Err(UpdateError::ImageMismatch);
+            }
+        }
+        Ok(())
+    }
+
+    // Looking at a region of flash, determine if there is a possible NXP
+    // image programmed. Return the length in bytes of the flash pages
+    // comprising the image including padding to fill to a page boundary.
+    fn flash_image_len(
+        &self,
+        span: &Range<u32>,
+    ) -> Result<usize, UpdateError> {
+        let buf = &mut [0u32; 1];
+        match indirect_flash_read(
+            &self.flash,
+            span.start + LENGTH_OFFSET as u32,
+            buf[..].as_bytes_mut(),
+        ) {
+            Ok(()) => {
+                if let Some(len) = round_up_to_flash_page(buf[0]) {
+                    // The minimum image size should be further constrained
+                    // but this is enough bytes for an NXP header and not
+                    // bigger than the flash slot.
+                    if len as usize <= span.len() && len >= HEADER_OFFSET {
+                        return Ok(len as usize);
+                    }
+                }
+                Err(UpdateError::BadLength)
+            }
+            Err(error) => Err(error),
+        }
+    }
+
+    fn cache_image_len(&self) -> Result<usize, UpdateError> {
+        let len = round_up_to_flash_page(
+            self.fw_cache[LENGTH_OFFSET / core::mem::size_of::<u32>()],
+        )
+        .ok_or(UpdateError::BadLength)?;
+
+        if len as usize > self.fw_cache.as_bytes().len() || len < HEADER_OFFSET
+        {
+            return Err(UpdateError::BadLength);
+        }
+        Ok(len as usize)
+    }
+
+    fn read_flash_image_to_cache(
+        &mut self,
+        span: Range<u32>,
+    ) -> Result<usize, UpdateError> {
+        // Returns error if flash page is erased.
+        let staged = image_range(RotComponent::Stage0, SlotId::B);
+        let len = self.flash_image_len(&staged.0)?;
+        if len as u32 > span.end || len > self.fw_cache.as_bytes().len() {
+            return Err(UpdateError::BadLength); // XXX unique error
+        }
+        indirect_flash_read(
+            &self.flash,
+            span.start,
+            self.fw_cache[0..len / core::mem::size_of::<u32>()].as_bytes_mut(),
+        )?;
+        Ok(len)
+    }
+
+    // XXX This API needs to be sorted out.
+    // The lower level routines are being replaced in PR 1500
+    fn write_cache_to_flash(
+        &mut self,
+        component: RotComponent,
+        slot: SlotId,
+    ) -> Result<(), UpdateError> {
+        let clen = self.cache_image_len()?;
+        if clen % BYTES_PER_FLASH_PAGE != 0 {
+            return Err(UpdateError::BadLength); // XXX need unique error
```

Comment:
Still need the unique error?

Reply:
I decided not. If some length in the image is invalid its not hard to figure out which one.